californium-core/src/main/java/org/eclipse/californium/core/network/stack/TcpBlockwiseLayer.java

```diff
@@ -0,0 +1,523 @@
+import org.eclipse.californium.core.network.config.NetworkConfig.Keys;
+
+/**
+ * Blockwise layer for TCP. Extended mainly for BERT option. Capable of handling
+ * incoming blocks with SZX = 7.
+ * 
+ * @author Pratheek Rai.
+ *
+ */
+public class TcpBlockwiseLayer extends BlockwiseLayer {
+
+	/**
+	 * logger.
+	 */
+	private static final Logger LOGGER = Logger.getLogger(TcpBlockwiseLayer.class.getName());
+
+	/**
+	 * BERT SZX constant.
+	 */
+	private final static int BERT_SZX = 7;
+	
+	/**
+	 * Internal block size for BERT. i.e internally BERT is handled similar to
+	 * SZX = 6.
+	 */
+	private final static int BERT_INT_BLOCK_SIZE = 1024;
+
+	/**
+	 * The number of 1024 Bytes packets to be sent as a block. Size of 1 BERT
+	 * block = bertStepSize * 1024 Bytes
+	 */
+	private int bertStepSize;
+
+	/**
+	 * Flag for BERT option.This flag is only for sending messages with BERT.
+	 * This class can handle incoming BERT messages irrespective of this flag.
+	 */
+	private boolean isBertEnabled;
+
+	/**
+	 * Constructor 
+	 * @param config - network configuration.
+	 */
+	public TcpBlockwiseLayer(NetworkConfig config) {
+		super(config);
+		bertStepSize = config.getInt(Keys.TCP_NUMBER_OF_BULK_BLOCKS, 1);
+		if (bertStepSize > 1) {
+			isBertEnabled = true;
+		}
+	}
+
+	/**
+	 * Handles inbound blockwise upload from peer. Handling is done here only
+	 * for SZX = 7 all other handling is done in the BlockwikseLayer.
+	 */
+	protected void handleInboundBlockwiseUpload(final Exchange exchange, final Request request) {
+
+		if (requestExceedsMaxBodySize(request)) {
+			// allow super class to handle this error.
+			super.handleInboundBlockwiseUpload(exchange, request);
+		} else {
+
+			BlockOption block1 = request.getOptions().getBlock1();
+
+			if ((block1 != null) && (block1.getSzx() == BERT_SZX)) {
+				KeyUri key = getKey(exchange, request);
+				Block1BlockwiseStatus status = getInboundBlock1Status(key, exchange, request);
+				
+				if (block1.getNum() == 0 && status.getCurrentNum() > 0) {
+					status = resetInboundBlock1Status(key, exchange, request);
+				}
+
+				if (block1.getNum() != status.getCurrentNum()) {
+					// ERROR, wrong number, Incomplete
+					LOGGER.log(Level.WARNING,
+							"peer sent wrong block, expected no. {0} but got {1}. Responding with 4.08 (Request Entity Incomplete)",
+							new Object[] { status.getCurrentNum(), block1.getNum() });
+
+					sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_INCOMPLETE,
+							"wrong block number");
+
+				} else if (!status.hasContentFormat(request.getOptions().getContentFormat())) {
+
+					sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_INCOMPLETE,
+							"unexpected Content-Format");
+
+				} else if (!status.addBlock(request.getPayload())) {
+					sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_TOO_LARGE,
+							"body exceeded expected size " + status.getBufferSize());
+				}
+
```

Comment:
Ok, now I see and think `} else {` is missing.
Without, the "errors" above will fall into the "success" below.